keps/sig-node/5328-node-capabilities/README.md

```diff
@@ -0,0 +1,821 @@
+
+2. Node Labels and Node Selectors/Affinity
+
+    Primarily used to **attract** specific nodes for pods based on the node's characteristics. By applying specific labels to nodes (reflecting Kubelet features, OS version, etc.), we can enable pods to use selectors or affinity to ensure they run on specific nodes.
+
+    **Standard usage pattern:**
+*   Cluster Administrators apply specific Labels to nodes to indicate the presence of certain features. This involves applying descriptive labels (Eg:_ kubelet.config.k8s.io/some-alpha-feature=enabled_, _node.kubernetes.io/gvisor-enabled=true_)
+*   Optionally for well defined features like[ RunTimeClass](https://kubernetes.io/docs/concepts/containers/runtime-class/), administrators may create abstractions (RuntimeClass Object) that bundle these feature requirements with a node selector targeting the corresponding node labels (Eg: spec._nodeSelector: { "node.kubernetes.io/runtime.gvisor": "true" })_
+*   Developers specify their workload's dependency on these features in the PodSpec either directly (spec.nodeSelector) or other mechanisms (eg: spec.runtimeClassName), enabling the scheduler to match them to capable nodes.
+
+**Drawbacks**:
+
+1. Operational Overhead: Cluster administrators have to add the necessary taints/labels to nodes as indirect signals of features and resources, and workloads should use corresponding tolerations/selectors. This needs to be done manually or automations built (webhooks, controllers) to handle standard usage patterns. 
+2. Scheduling constraints are encoded indirectly rather than being implicitly understood by kube-scheduler. Discoverability 
+3. Incorrect configurations can lead to scheduling failures or suboptimal placements.
+
+## Proposal
+
+This proposal introduces a new field `NodeCapabilities `to` NodeSpec.NodeStatus `to expose information which the kube-scheduler and/or the admission controller would use to make more informed decisions.` `Kubelet is primarily responsible for discovering, consolidating, and reporting `NodeCapabilities` to the API server. The data exposed as NodeCapabilities falls into these categories:
+
+1. **Kubelet Configuration:** Includes relevant kubelet feature gates and configurations. A single logical capability can represent multiple underlying settings.
+2. **Runtime Configurations:** Covers scheduler-relevant runtime details like supported handlers and features, obtained via the existing [Status RPC](https://github.com/kubernetes/cri-api/blob/79a12c13b6e0a0e049d166d72b0f352057d9e41b/pkg/apis/runtime/v1/api.proto#L117) between Kubelet and CRI.
+3. **Kernel Configurations:** Details kernel support necessary for certain pod features, which may be optional or available only in newer kernel versions.
+
+**NodeCapabilities Requirements:**
+
+1. Must be derived from static configuration.
```

Comment:
This rules out any capabilities derived from plugins that can be registered after startup. That seems OK for an MVP, but are we open to relaxing this requirement in the future?

Reply:
> The key challenge would be to ensure that the capabilities are published before the workload pods are scheduled.

I don't think that would be a problem with the way capabilities are currently defined? In other words, the only way the scheduler uses capabilities is by saying that a certain capability is required. If plugin capabilities are strictly additive, it just means that workloads that require those plugin capabilities wouldn't be able to be scheduled before they're added.

Reply:
Thanks. Makes sense. Temporary scheduling delay should be fine and even safe. The proposed API (node.status.capabilities) itself should be broad enough to handle plugin-derived capabilities. I'll update the KEP to include this as a goal for Beta, along with the necessary considerations.